### DISCUSSION TITLE
Fill enclosed islands in radius map + small shiftFill.bat correction (snibgo)

### DISCUSSION THREAD

**Discussion by ranknfyle** [ImageMagick/ImageMagick#7856] [Help] [answered]
ImageMagick version
7.1.1-43-Q8-x64

Operating system, version and so on
Win 10

1. I use maxCircle.bat and it works fine. I'm dealing with a fixed radius, so my min and max radii (args %5 & %6) are the same. The radius map I get by setting %7 to "-auto-level -write ramap.png" looks correct: a map of areas where a circle would fit in. The areas are shown as islands surrounded by a white line (img: https://i.ibb.co/YhLk5KC/map.png). I'd like to fill in these islands with white. I tried shiftFill.bat in %PICTBAT%, setting the bg (%sfFLAT_COL) to "Black" in the script. shiftFill finishes without errors but the islands are unchanged in the output (not filled-in). Perhaps the reason is that the radius map bg is black, not transparent and shiftFill.bat needs transparency. Also, it takes a while. I know Python so I can do this with Python libs. But just in case: can the radius map fill in the islands (any color will do) to save the need to fill them later?

2. In shiftFill.bat the input and output files are written to a temp script. The output file (%OUTFILE) gets double backslashes "\\", so it's correctly read in. The input file (%INFILE) doesn't get them, so magic stops due to missing backslashes: (magick: unable to open image 'C:UsersXDocuments...')  etc. The input file should also get the 2-backslash treatment:

```
set INFILE2=!INFILE:\=\\!
echo %INFILE2% +write mpr:ORIG -compose DstOver >%TMP_SCR%

```

**Comment by snibgo** [comment, marked as answer]
> I'd like to fill in these islands with white.

The obvious method is to flood-fill from the top-left corner in any colour other than black or white, such as red. Then change black pixels into white. Then change red pixels into black.
```
magick map.png -fill Red -draw "color 0,0 floodfill" -fill White -opaque Black -fill Black -opaque Red map2.png
```
This method assumes the pixel at top-left can't be within an island, or part of the white border around an island. If that can't be guaranteed, then a different method should be used, perhaps using "-connected-components" to find the largest black area.

> The input file should also get the 2-backslash treatment.

I agree. Thanks for the correction. The same applies to any of my BAT scripts that use an IM script to read input files. I'll do the necessary searching and correcting.

**Reply by ranknfyle** [reply]
I've just encountered a problem with this method: an island within an island. The smaller inner island is filled, too, but it shouldn't be [link to image](https://postimg.cc/nCBqky7q) (the fill color is really white, but I made it gray so we can see the original white lines). The white line surrounding the inner island doesn't have any gaps, so gaps are not an issue. Is there a way to deal with this island-within-an-island problem? Or, at least, auto-detect and flag such a case (one delineated area surrounded by another one), so I can fix it manually?

PS. Just thinking out loud: one way to detect (not fix) such cases would be to take the final image (islands filled white) and multiply it by the inverse (-negate) of the original image with outlines. The white-filled islands will shrink slightly (because of the now-black outline) but, more importantly, any "inner-islands" will now show as black outlines within white areas. So now I can "search" for any spots where a black pixel has white pixels on both sides (up/down or left-right). I'll try to formalize this later.

**Reply by snibgo** [reply]
To help our understanding, please show an actual input image, and the output you want (perhaps created with Gimp).

**Comment by ranknfyle** [comment]
> flood-fill from the top-left corner

Thanks, it works great. If an island is in the top-left (very rare), I'll alter it manually.

**Reply by fmw42** [reply]
Just add a border around the image of 1 pixel, then flood fill from the new corner at 0,0.  See -bordercolor and -border.

**Comment by snibgo** [comment]
As fmw42 says, "-bordercolor Black -border 1" before the floodfill at 0,0 , and "-shave 1" after. This will fix any images that have the white island-border at top-left.
